### ISSUE TITLE
Support for multi subtitles display

### ISSUE THREAD

**Issue by mailwork** [UniversalMediaServer/UniversalMediaServer#1214]
I've used to watch movies with english and my mother language at same time on KMPlayer.
While UMS packs multiple text tracks to video container so it shouldn't be any problem to display  two subs at same time and I guess it will be useful and interesting for other people too.

**Comment by Sami32** [comment]
As i don't even know if some TV support double subtitles displaying, i guess that the way to make it work would be to burn them.
That's mean that more CPU power will be need.

As @valib is the more subtitled oriented among us, i'm sure he have a better overview on the subject.

Overlay will be better, but i never looked at it, and don't know if it's even possible to correctly overlay 2 subtitles with FFmpeg.

**Comment by Nadahar** [comment]
I really can't understand that this is useful at all - is it for learning purposes? It will block a lot of the picture however you put it.

**Comment by Sami32** [comment]
@Nadahar From what i understood, it's for language learning purpose. That's make sens to me.
I guess that other use, as 2 different nationality people (perhaps wife or grand-parents, friends...) viewing a movie together but no one understand fully their adoptive (or VO) language.

In my case, i use the subtitle and she get the audio V.O., or the opposite, depending of the movie.

**Comment by mailwork** [comment]
@Nadahar Yes, that's perfect for learning purposes! Also there are always some missing and mistakes in translations.

**Comment by Nadahar** [comment]
@mailwork I almost always use subtitles myself, but I don't experience that as a big problem. A lot of the content has English audio though, which I understand reasonably well - but I also prefer subtitles because that means that I don't have to have the audio volume high enough to hear what they say in quiet or noisy circumstances - because that usually means that it will be way too loud at other times.

**Comment by Sami32** [comment]
Or the users accept to get the subtitles "eat" around half of their screen or some renderers are able to display them in the black band, when the movies are in the cinema format or resized by choice.

**Comment by Nadahar** [comment]
@Sami32 I haven't said it doesn't have its uses, I just said that I use subtitles a lot (almost always) and I don't see a lot of lacking translations even though sometimes they are way off. That's usually because the translator apparantly knew nothing about the subject being talked about. But, I guess subtitle quality can vary as well. I almost always understand enough of the audio to "fill in" what's missing from the subtitles though.

I'm not saying we shouldn't implement this, but I think it has a pretty narrow use case. It would probably be most interesting in homes with mixed nationalities and little overlap in language skills. I suspect it could be tricky to get consistent though, it would either require transcoding or that we wrote some kind of "parser" which could combine the two into one subtitle screen. There's the character encoding and subtitle formats to deal with as well, so generating such a combined subtitle stream could be complicated.

**Comment by Sami32** [comment]
@Nadahar Not to burn a movie with 2 subtitles, as far as i see it, and they will not use more CPU than for 1 subtitle anyway. So that's not an issue in that case.

I don't see the characters encoding as a issue, as `-sub_charenc` could be used, not even with different kind of text based subtitles.
The overlay positioning should more tricky to find, but reachable i guess, perhaps with test and try.

**Comment by Sami32** [comment]
Some SRT have the X1 X2 Y1 Y2 tags that can make it even more easy, as with SSA subtitles.

For the others text based ones, `force_style=Alignment` can be used:
`force_style='FontName=Arial,FontSize=13,Alignment=10'" -vcodec libx264 ...`

The more easy way to be implemented, now i remember some videos having 2 subtitles, one upstair and one downstair. More readable as well i think. `Alignment` set it topup by default, and without it's topdown by default.

**Comment by Nadahar** [comment]
@Sami32 You're talking about something completely different than I was. I understand that transcoding is one possibility, what I was talking about was to do it without forcing transcoding. It would be possible to combine two (external) subtitles into one, but it would be complicated. That's where character encoding and the other things I mentioned come into play.

That said, I don't think I've ever watched a movie with my grand parents.

**Comment by Sami32** [comment]
@nadahar Crazy individual world, where family become sparse...

I was talking about the 2 possibilities since the beginning. Remember, no problems only solutions ;-)
Your subtitles mixing, through possible, is not a good approach from my point of that i already explained, ask more work and less visibility, IMO.

**Comment by Nadahar** [comment]
@Sami32 I still don't think you understand what I mean. It wouldn't be any less visible. The natural thing to do would have been to generate an ASS file as the combined subtitle, and then you could place them as you'd like as far as I remember. But, by all means, forget about it - I didn't say I thought it was a good approach - because all the conversion would be a nightmare to get bug free.

**Comment by Sami32** [comment]
@Nadahar For me your approach is wrong.

I guess you misunderstood the FFmpeg command line, but as you say talking or writing about the subject will not make it happen.

**Comment by Nadahar** [comment]
@Sami32 I was talking about NOT transcoding - ergo no FFmpeg. There would be no reason to do something like that if you were still going to transcode it.

**Comment by Sami32** [comment]
@Nadahar Yes, and for me it's a wrong approach. Too much work for saving CPU from transcoding...

In a perfect work our TVs will be able to read natively 2 subtitles at the same time ;-)

**Comment by Nadahar** [comment]
@Sami32 Saving the CPU isn't the only reason to avoid transcoding. There's delay, sync, quality loss, loss of multiple audio tracks, loss of seek/fast forward to name a few. I really don't want to transcode unless it's so way around it, and I suspect I'm not the only one.

**Comment by Sami32** [comment]
@Nadahar I guess i said remuxing, that perfectly suffisant to me for 2 subtitles, IMO.

**Comment by Nadahar** [comment]
@Sami32 Remuxing still have many of the same disadvantages as transcoding, even though quality loss is excluded.

**Comment by Sami32** [comment]
SSA are less widely supported than SRT. But FFmpeg can convert, concatenate, and display them.
What a beautiful tool.

**Comment by Sami32** [comment]
@mailwork How do you would like them displayed ?
1) One on the upper part and the other on the bottom part, overlayed on the video ?
or
2) Subtitles set on  upper and lower black bands, with a resized down video ?
